components/ActionSheet/src/MDCActionSheetController.m

```diff
@@ -0,0 +1,331 @@
+  [_tableView.tableView setNeedsLayout];
+  initialLayout = false;
+  [self.view setNeedsLayout];
+}
+
+- (NSArray<MDCActionSheetAction *> *)actions {
+  return _tableView.actions;
+}
+
+- (void)viewDidLoad {
+  [super viewDidLoad];
+  
+  if (self.header == nil) {
+    self.header = [self headerView];
+  }
+
+  [self.view addSubview:_tableView.tableView];
+  [self.view addSubview:self.header];
+}
+
+- (void)viewWillLayoutSubviews {
+  [super viewWillLayoutSubviews];
+
+  if (initialLayout == false) {
+    [self firstLayout];
+  }
+}
+
+-(void)firstLayout {
+  [self.header setNeedsLayout];
+  [self.header layoutIfNeeded];
+
+  CGFloat width = CGRectGetWidth(self.view.bounds);
+  CGFloat height = CGRectGetHeight(self.header.frame) + [_tableView tableHeightForWidth:width];
+  CGRect tableFrame = _tableView.tableView.frame;
+  tableFrame.origin.y = CGRectGetHeight(self.header.frame);
+  _tableView.tableView.frame = tableFrame;
+  self.preferredContentSize = CGSizeMake(width, height);
+  initialLayout = true;
+}
+
+- (void)viewWillAppear:(BOOL)animated {
+  [super viewWillAppear:animated];
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+  self.mdc_bottomSheetPresentationController.delegate = self;
+#pragma clang diagnostic pop
+
+  self.mdc_bottomSheetPresentationController.dismissOnBackgroundTap =
+      _transitionController.dismissOnBackgroundTap;
+  [self.view layoutIfNeeded];
+}
+
+- (BOOL)accessibilityPerformEscape {
+  if (!self.dismissOnBackgroundTap) {
+    return NO;
+  }
+  [self dismissViewControllerAnimated:YES completion:nil];
+  return YES;
+}
+
+- (void)preferredContentSizeDidChangeForChildContentContainer:(id<UIContentContainer>)container {
+  [super preferredContentSizeDidChangeForChildContentContainer:container];
+  [self.presentationController
+      preferredContentSizeDidChangeForChildContentContainer:self];
+}
+
+- (UIScrollView *)trackingScrollView {
+  return _transitionController.trackingScrollView;
+}
+
+- (void)setTrackingScrollView:(UIScrollView *)trackingScrollView {
+  _transitionController.trackingScrollView = trackingScrollView;
+}
+
+- (BOOL)dismissOnBackgroundTap {
+  return _transitionController.dismissOnBackgroundTap;
+}
+
+-(void)setDismissOnBackgroundTap:(BOOL)dismissOnBackgroundTap {
+  _transitionController.dismissOnBackgroundTap = dismissOnBackgroundTap;
+  self.mdc_bottomSheetPresentationController.dismissOnBackgroundTap = dismissOnBackgroundTap;
+}
+
+- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:
+    (id<UIViewControllerTransitionCoordinator>)coordinator {
+  [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
+  CGRect frame = self.view.frame;
+  frame.size = size;
+  self.view.frame = frame;
+  CGRect headerFrame = self.header.frame;
+  headerFrame.size.width = size.width;
+  self.header.frame = headerFrame;
+  [self.header setNeedsLayout];
+  [self.header layoutIfNeeded];
```

Comment:
`layoutIfNeeded` should rarely be called directly because it forces an immediate layout rather than allowing UIKit to do the layout as a batch at the end of the current run loop.

If you do need to invoke `layoutIfNeeded`, please document why.

Reply:
This causes an extra layout call to the header object but since we are using the header's height when setting the preferredContentSize if we don't make this call then the header's height is 0.